### PR TITLE
Add libtiff

### PR DESCRIPTION
**Pull request by ocefpaf**
Trying to solve `jpeg`/`gdal`/`fiona` issue.

### CODE REVIEW

**Comment by ocefpaf** [comment]
Master is not happy:

```shell
Repository created, please edit conda-forge.yml to configure the upload channels
and afterwards call 'conda smithy register-github'
Traceback (most recent call last):
  File ".CI/create_feedstocks.py", line 107, in <module>
    if repo_exists('conda-forge', os.path.basename(feedstock_dir)):
  File ".CI/create_feedstocks.py", line 55, in repo_exists
    org = gh.get_organization(organization)
  File "/Users/travis/miniconda/lib/python3.5/site-packages/github/MainClass.py", line 188, in get_organization
    "/orgs/" + login
  File "/Users/travis/miniconda/lib/python3.5/site-packages/github/Requester.py", line 171, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, cnx))
  File "/Users/travis/miniconda/lib/python3.5/site-packages/github/Requester.py", line 179, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.GithubException: 403 {'documentation_url': 'https://developer.github.com/v3/#rate-limiting', 'message': 'API rate limit exceeded for conda-forge-admin.'}
```

**Comment by jakirkham** [comment]
Might want to be careful there. Deployment is failing.

**Comment by jakirkham** [comment]
Yeah, it's been like that for a little bit. I tried restarting a couple times, but I think we need to wait for a bit.

**Comment by ocefpaf** [comment]
I won't merge this. Just testing the build. (In fact I am going to :zzz: now!)

**Comment by jakirkham** [comment]
Ok, sure.

**Comment by jakirkham** [comment]
Cleared `master`. Think we just had too many GitHub API requests too close together. Breaking the recipes into smaller groups got them through without any trouble.

As far as the failure, it sounds like a case for `DYLD_FALLBACK_LIBRARY_PATH`. Unfortunately, not a superhero, but it may just fix these test failures.

**Comment by ocefpaf** [comment]
@jakirkham thanks for clearing `master`. I rebased and updated the recipe. Can you take a look and merge?

File: recipes/jbig/build.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# Patch (support old gcc).
+sed -i.bak 's/-Wno-unused-result//g' Makefile
```
Comment (jakirkham) [inline recipes/jbig/build.sh]: Should be fine as we are just removing a disable warning option.

File: recipes/jbig/Makefile.patch [hunk 2]
```diff
@@ -0,0 +1,11 @@
+--- Makefile	2014-03-27 14:47:15.000000000 -0400
++++ Makefile	2015-11-06 14:14:48.000000000 -0500
+@@ -4,7 +4,7 @@
+ CC = gcc
+ 
+ # Options for the compiler: A high optimization level is suggested
+-CFLAGS = -O2 -W -Wno-unused-result
++CFLAGS = -O2 -W -fPIC
```
Comment (jakirkham) [inline recipes/jbig/Makefile.patch]: So, the `sed` command takes care of part of this. Why is it we `-fPIC` here? Couldn't we just set the `CFLAGS` in advance?

File: recipes/jbig/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+test:
+    commands:
+        - test -f ${PREFIX}/lib/libjbig.a  # [unix]
+        - test -f ${PREFIX}/lib/libjbig85.a  # [unix]
+        - conda inspect linkages -n _test jbig  # [linux]
+
+about:
+    home: https://www.cl.cam.ac.uk/~mgk25/jbigkit
+    license: GPL-2.0+
```
Comment (jakirkham) [inline recipes/jbig/meta.yaml]: What's the `+` here for?

File: recipes/libtiff/build.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            --with-zlib-lib-dir="${PREFIX}/lib" \
+            --with-jpeg-include-dir="${PREFIX}/include" \
+            --with-jpeg-lib-dir="${PREFIX}/lib" \
+            --with-jbig-include-dir="${PREFIX}/include" \
+            --with-jbig-lib-dir="${PREFIX}/lib" \
+            --with-lzma-include-dir="${PREFIX}/include" \
+            --with-lzma-lib-dir="${PREFIX}/lib" \
+            --bindir="${TIFF_BIN}" \
+            --datarootdir="${TIFF_SHARE}" \
+            --docdir="${TIFF_DOC}" \
```
Comment (jakirkham) [inline recipes/libtiff/build.sh]: Is there some way to disable this sort of stuff (docs and data)?

File: recipes/libtiff/build.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+./configure --prefix="${PREFIX}" \
+            --with-zlib-include-dir="${PREFIX}/include" \
+            --with-zlib-lib-dir="${PREFIX}/lib" \
+            --with-jpeg-include-dir="${PREFIX}/include" \
+            --with-jpeg-lib-dir="${PREFIX}/lib" \
+            --with-jbig-include-dir="${PREFIX}/include" \
+            --with-jbig-lib-dir="${PREFIX}/lib" \
+            --with-lzma-include-dir="${PREFIX}/include" \
+            --with-lzma-lib-dir="${PREFIX}/lib" \
+            --bindir="${TIFF_BIN}" \
```
Comment (jakirkham) [inline recipes/libtiff/build.sh]: Normally, this gets set correctly when setting `--prefix`. Was that not the case?

File: recipes/libtiff/build.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+if [[ $(uname) == Darwin ]]; then
+  export LIBRARY_SEARCH_VAR=DYLD_FALLBACK_LIBRARY_PATH
+elif [[ $(uname) == Linux ]]; then
+  export LIBRARY_SEARCH_VAR=LD_LIBRARY_PATH
+fi
+
+# Install non-lib artifacts to these special directories so we can cleanly delete them later.
+TIFF_BIN="${PREFIX}/tiff-bin"
```
Comment (jakirkham) [inline recipes/libtiff/build.sh]: So, wait, what is getting placed in this directory?

File: recipes/libtiff/build.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+            --with-jpeg-lib-dir="${PREFIX}/lib" \
+            --with-jbig-include-dir="${PREFIX}/include" \
+            --with-jbig-lib-dir="${PREFIX}/lib" \
+            --with-lzma-include-dir="${PREFIX}/include" \
+            --with-lzma-lib-dir="${PREFIX}/lib" \
+            --bindir="${TIFF_BIN}" \
+            --datarootdir="${TIFF_SHARE}" \
+            --docdir="${TIFF_DOC}" \
+
+make -j"${CPU_COUNT}"
```
Comment (jakirkham) [inline recipes/libtiff/build.sh]: Probably should drop the parallel option just to be safe. Some CIs are not great at telling us what level of parallelism we actually have versus whether we are sitting on a node with other builds and sharing these cores.

File: recipes/libtiff/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+package:
+    name: libtiff
+    version: {{ version }}
+
+source:
+    fn: tiff-{{ version }}.tar.gz
+    url: ftp://ftp.remotesensing.org/pub/libtiff/tiff-{{ version }}.tar.gz
+    md5: d1d2e940dea0b5ad435f21f03d96dd72
+    patches:
+        - def_snprintf.patch  # [win]
```
Comment (jakirkham) [inline recipes/libtiff/meta.yaml]: Where did we get this patch from?

File: recipes/jbig/Makefile.patch (hunk 2 above)
Comment (msarahan) [inline recipes/jbig/Makefile.patch]: also, it's really, really bad to set CFLAGS without respecting any existing setting.  Please always do

    CFLAGS = $CFLAGS <+ additions>

For example, one of the things this breaks is the multilib 32-bit building.

File: recipes/jbig/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    name: jbig
+    version: {{ version }}
+
+source:
+    fn: jbigkit-{{ version }}.tar.gz
+    url: https://www.cl.cam.ac.uk/~mgk25/jbigkit/download/jbigkit-{{ version }}.tar.gz
+    md5: ebcf09bed9f14d7fa188d3bd57349522
+    patches:
+        # Remove -Wno-unused-result (for compatibility with old gcc), and add -fPIC (to allow shared libraries to use this build).
+        - Makefile.patch
```
Comment (jakirkham) [inline recipes/jbig/meta.yaml]: With notes above, will this patch still be needed?

File: recipes/jbig/Makefile.patch (hunk 2 above)
Comment (ocefpaf) [inline recipes/jbig/Makefile.patch]: Yeah. Those were in the [original recipe](https://github.com/conda/conda-recipes/tree/0dabe789218ef58df023154dfc6482791b336451/jbig). I am not sure we need that patch though. I just removed it and let's see what happens.

File: recipes/jbig/meta.yaml (hunk 3 above)
Comment (ocefpaf) [inline recipes/jbig/meta.yaml]: That is an old syntax for GPL2 or later.

File: recipes/jbig/Makefile.patch (hunk 2 above)
Comment (jakirkham) [inline recipes/jbig/Makefile.patch]: I added my question on the [commit]( https://github.com/conda/conda-recipes/commit/5609b2a1b4bd7fd8fe90f01747a48e7ece53b947#commitcomment-17132447 ) that added the patch too.

File: recipes/jbig/meta.yaml (hunk 9 above)
Comment (ocefpaf) [inline recipes/jbig/meta.yaml]: I just test exporting  `CFLAGS="$CFLAGS -fPIC"`, but that does does not work. We do need the patch. (Or someone smarter than me.)

File: recipes/libtiff/build.sh (hunk 4 above)
Comment (ocefpaf) [inline recipes/libtiff/build.sh]: Done.

File: recipes/libtiff/build.sh (hunk 7 above)
Comment (ocefpaf) [inline recipes/libtiff/build.sh]: Done.

File: recipes/libtiff/build.sh (hunk 5 above)
Comment (ocefpaf) [inline recipes/libtiff/build.sh]: Not sure. I just copied the recipe. Let me test here locally and I will get back to you.

File: recipes/libtiff/build.sh (hunk 6 above)
Comment (ocefpaf) [inline recipes/libtiff/build.sh]: Not sure. I removed all that and let's see how it goes.

File: recipes/libtiff/meta.yaml (hunk 8 above)
Comment (ocefpaf) [inline recipes/libtiff/meta.yaml]: From the original recipe in conda-recipes, but I see that is a similar fix to what we did to gdal.

File: recipes/jbig/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/jbig/meta.yaml]: Maybe drop the `+`. Most people familiar with the GPL know that they have such a clause.

**Comment by ocefpaf** [comment]
@jakirkham and @msarahan I addressed all the review comments. (Thanks!)

**Comment by stuarteberg** [comment]
Hey guys, I see you copied from `conda-recipes/libtiff` here.  I regret adding `jbig` support to that recipe.  I don't know anyone who actually uses it (apparently it was written for, um, fax machines).  At the time, I just figured "Hey, libtiff supports this other compression scheme, so why not add support for it?"

But since it adds almost-zero value for data scientists and NON-zero overhead to the build maintenance effort, you should consider just removing it from the build entirely.  I bet no one will miss it.  From the discussion above, I can see it's already wasted some of your time.  Apologies for that.  Maybe you can save future maintainers the trouble.

Just my two cents.

**Comment by jakirkham** [comment]
> ...apparently it was written for, um, fax machines...

Hilarious. :laughing: Reminds me of 30 Rock.

> Beepers are going to make a comeback, technology is cyclical.

-- The Pager King, Dennis Duffy

**Comment by ocefpaf** [comment]
Thanks @stuarteberg! I will remove it.

File: recipes/libtiff/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        - conda inspect linkages -n _test libtiff  # [linux]
+
+about:
+    home: http://www.remotesensing.org/libtiff/
+    license: HPND
+    summary: Support for the Tag Image File Format (TIFF)
+
+extra:
+    recipe-maintainers:
+        - ocefpaf
```
Comment (ocefpaf) [inline recipes/libtiff/meta.yaml]: @stuarteberg since you created this recipe can I add you as a maintainer :wink:?

**Comment by ocefpaf** [comment]
@jakirkham I am merging this because I want to fix `gdal`. If you have any further comments/suggestions I will be happy to implement them in the feedstock.

File: recipes/libtiff/meta.yaml (hunk 10 above)
Comment (stuarteberg) [inline recipes/libtiff/meta.yaml]: Sure.

File: recipes/libtiff/meta.yaml (hunk 8 above)
Comment (jakirkham) [inline recipes/libtiff/meta.yaml]: The closest thing I can find about needing a patch like this a [thread]( http://comments.gmane.org/gmane.comp.video.tiff/6889 ) that doesn't include the patch. I was hoping I would find exactly where this patch came from, but no such luck. I know this is tedious, sorry in advance, but would you mind submitting it upstream. You could probably even just drop it into that thread. Here is their [mailing list]( http://lists.maptools.org/mailman/listinfo/tiff ) sign up page.

Comment (jakirkham) [inline recipes/libtiff/meta.yaml]: Sorry, I read that wrong. The user is effectively explaining the patch. Maybe just add a link to that thread here instead.

Comment (ocefpaf) [inline recipes/libtiff/meta.yaml]: Will do.

**Comment by jakirkham** [comment]
That's fine. Just adding a link to that thread above the patch. Otherwise everything else is fine.

File: recipes/libtiff/meta.yaml (hunk 8 above)
Comment (msarahan) [inline recipes/libtiff/meta.yaml]: It seems that I wrote it: https://github.com/conda/conda-recipes/blob/0dabe789218ef58df023154dfc6482791b336451/libtiff/def_snprintf.patch .  I vaguely recall something about it.  I guess if you're worried about it, leave it out and see what happens.  I think you run into either duplicate definition of snprintf or lack of it, depending on the compiler you're using.

Comment (ocefpaf) [inline recipes/libtiff/meta.yaml]: I haven't tested it but I believe we need it. I saw the same patch in other places in similar situations. (I will try to get my hands on a Windows machine to confirm it though.)

**Comment by jakirkham** [comment]
@stuarteberg, since this is your first time being added as a maintainer at conda-forge, I am going to explain what is about to happen. When this recipe gets merged, it will be converted to a feedstock (repo) (see [here]( https://github.com/conda-forge )) that contains the recipe in a directory so named and some CI files used for automated build and deployment of packages to the [`conda-forge` channel]( http://anaconda.org/conda-forge/ ). This will trigger its first build and release. After this happens a separate CI script will be triggered that will add this to our webpage listing of packages, add it to the [feedstocks repo]( https://github.com/conda-forge/feedstocks ) (like conda-recipes with git submodules), and then invite you to join conda-forge. After accepting the invite, you will be added to a team called `libtiff`. This gives you commit rights over the feedstock (repo) and control over its respective CIs. It will also automatically subscribe you to watch that feedstock (repo). In the future, simply having your GitHub handle added by yourself or by someone (with your permission) will be all that is necessary to add you to other feedstocks.

If you have questions or concerns, feel free to let us know. Welcome to conda-forge.

File: recipes/libtiff/meta.yaml (hunk 8 above)
Comment (jakirkham) [inline recipes/libtiff/meta.yaml]: Yeah, I'm pretty sure it is necessary was just trying to track down if this has reached upstream yet. Sounds like there are at least aware of the problem and we seem to be using the same patch others are.

**Pull request merged**